opm/simulators/linalg/MatrixMarketUtils.hpp

```diff
@@ -0,0 +1,101 @@
+/*
+  Copyright 2019 SINTEF Digital, Mathematics and Cybernetics.
+  Copyright 2019 Dr. Blatt - HPC-Simulation-Software & Services.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_MATRIXMARKETUTILS_HEADER_INCLUDED
+#define OPM_MATRIXMARKETUTILS_HEADER_INCLUDED
+
+#include <dune/istl/matrixmarket.hh>
+
+#include <cstddef>
+
+namespace Dune
+{
+
+namespace MatrixMarketImpl
+{
+    template <typename T, typename A, int brows, int bcols, typename D>
+    void makeSparseEntries(Dune::BCRSMatrix<Dune::FieldMatrix<T, brows, bcols>, A>& matrix,
+                           std::vector<int>& i,
+                           std::vector<int>& j,
+                           std::vector<double>& val,
+                           const D&)
+    {
+        // addapted from readSparseEntries
+        typedef Dune::BCRSMatrix<Dune::FieldMatrix<T, brows, bcols>, A> Matrix;
+        std::vector<std::set<IndexData<D>>> rows(matrix.N() * brows);
+        for (std::size_t kk = 0; kk < i.size(); ++kk) {
```

Comment:
On a second read, I kind of get it. This has nothing to do with matrixmarket besides the fact that you are misusing the dune-istl code for an (inefficient) matrix setup.

BTW there is no need for rows as you can pass this information directly to CreateIterator. You are wasting memory here.